nilearn/decomposition/dict_learning.py

```diff
@@ -0,0 +1,201 @@
+        This parameter is passed to image.resample_img. Please see the
+        related documentation for details.
+
+    low_pass: None or float, optional
+        This parameter is passed to signal.clean. Please see the related
+        documentation for details
+
+    high_pass: None or float, optional
+        This parameter is passed to signal.clean. Please see the related
+        documentation for details
+
+    t_r: float, optional
+        This parameter is passed to signal.clean. Please see the related
+        documentation for details
+
+    memory: instance of joblib.Memory or string
+        Used to cache the masking process.
+        By default, no caching is done. If a string is given, it is the
+        path to the caching directory.
+
+    memory_level: integer, optional
+        Rough estimator of the amount of memory used by caching. Higher value
+        means more memory for caching.
+
+    n_jobs: integer, optional, default=1
+        The number of CPUs to use to do the computation. -1 means
+        'all CPUs', -2 'all CPUs but one', and so on.
+
+    verbose: integer, optional
+        Indicate the level of verbosity. By default, nothing is printed
+
+    References
+    ----------
+
+    """
+
+    def __init__(self, mask=None, n_components=20,
+                 smoothing_fwhm=6,
+                 standardize=True,
+                 random_state=0,
+                 target_affine=None, target_shape=None,
+                 low_pass=None, high_pass=None, t_r=None,
+                 alpha=1,
+                 n_iter=1000,
+                 # Common options
+                 memory=Memory(cachedir=None), memory_level=0,
+                 n_jobs=1, verbose=0,
+                 dict_init=None
+                 ):
+        SinglePCA.__init__(self,
+                           mask=mask, memory=memory, memory_level=memory_level,
+                           n_jobs=n_jobs, verbose=verbose,
+                           n_components=n_components, smoothing_fwhm=smoothing_fwhm,
+                           target_affine=target_affine, target_shape=target_shape,
+                           random_state=random_state, high_pass=high_pass, low_pass=low_pass,
+                           t_r=t_r,
+                           standardize=standardize,
+                           do_cca=False)
+        self.n_iter = n_iter
+        self.alpha = alpha
+        self.dict_init = dict_init
+
+    def _init_dict(self, imgs, y=None, confounds=None):
+        if self.dict_init is not None:
+            self.dict_init_ = self.dict_init
+        else:
+            # Perform a CanICA initialization using SinglePCA parameters
+            parameters = get_params(SinglePCA, self)
+            # Efficient CanICA requires CCA
+            parameters['do_cca'] = True
+            canica = CanICA(threshold=float(self.n_components), n_init=1,
+                            **parameters)
+            canica.fit(imgs, y=y, confounds=confounds)
```

Comment:
I've been thinking about your optimization. In some sklearn functions, there is a switch to allow multiple inputs. In [Spectral clustering](http://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html), you can specify `affinity='precomputed'` and directly give the affinity matrix in the fit.

You could do the same here for the subject pca.